cli/src/commands/extensions/mod.rs

```diff
@@ -0,0 +1,132 @@
+pub mod extension;
+
+use std::{collections::HashSet, convert::TryFrom, path::PathBuf};
+
+use crate::commands::{CommandResult, CommandValue, ExitCode};
+pub use extension::*;
+
+use anyhow::Result;
+use clap::{arg, ArgMatches, Command, ValueHint};
+use log::{error, warn};
+
+pub fn command<'a>() -> Command<'a> {
+    Command::new("extension")
+        .about("Run extensions")
+        .subcommand(
+            Command::new("add")
+                .about("Install extension")
+                .arg(arg!([PATH]).required(true).value_hint(ValueHint::DirPath)),
+        )
+        .subcommand(
+            Command::new("remove")
+                .about("Uninstall extension")
+                .arg(arg!([NAME]).required(true)),
+        )
+        .subcommand(Command::new("list").about("List installed extensions"))
+}
+
+/// Generate the subcommands for each extension.
+/// TODO add tests.
+pub fn extensions_subcommands(command: Command<'_>) -> Command<'_> {
+    let extensions = match installed_extensions() {
+        Ok(extensions) => extensions,
+        Err(e) => {
+            error!("Couldn't list extensions: {}", e);
+            return command;
+        }
+    };
+
+    let names = command
+        .get_subcommands()
+        .map(|n| n.get_name().to_string())
+        .collect::<HashSet<_>>();
+
+    extensions
+        .into_iter()
+        .filter(|ext| {
+            if names.contains(ext.name()) {
+                warn!(
+                    "{}: extension was filtered out due to name conflict",
+                    ext.name()
+                );
+                false
+            } else {
+                true
+            }
+        })
+        .fold(command, |command, ext| {
+            command.subcommand(Command::new(ext.name()))
+        })
+}
+
+/// Entry point for the `extensions` subcommand.
+pub async fn handle_extensions(matches: &ArgMatches) -> CommandResult {
+    if let Some(matches) = matches.subcommand_matches("add") {
+        handle_add_extension(matches.value_of_t("PATH")?).await
+    } else if let Some(matches) = matches.subcommand_matches("remove") {
+        handle_remove_extension(matches.value_of("NAME").unwrap()).await
+    } else {
+        // also covers the `list` case
+        handle_list_extensions().await
+    }
```

Comment:
Other parts of our code mention [switching to match expressions](https://github.com/phylum-dev/cli/blob/8f112df9adcfb8c06bd6465be84c190eb8fe33db/cli/src/bin/phylum.rs#L196) for subcommands. You could be the first to actually do it.

```suggestion
    match matches.subcommand() {
        Some(("add", matches)) => handle_add_extension(matches.value_of_t("PATH")?).await,
        Some(("remove", matches)) => handle_remove_extension(matches.value_of("NAME").unwrap()).await,
        Some(("list", _)) | None => handle_list_extensions().await,
        _ => unreachable!(),
    }
```

Reply:
I think the catch-all branch may actually be reachable in the `None` form, as we want to handle the no-subcommand case. I will commit this suggestion and then check.

Reply:
Notice that the last pattern is `Some(("list", _)) | None`

Reply:
Sorry, missed that!